### ISSUE TITLE
To unify the regularizer in keras layer and variable_scope.py?

### ISSUE THREAD

**Issue by wangsiyu** [tensorflow/tensorflow#29196]
**System information**
- TensorFlow version (you are using): master
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

**Will this change the current api? How?**
Don't change current api.

**Who will benefit with this feature?**
I think the regularizer should be declared with variables creation. It is reasonable to put them in variable_scope.py after one atomic variable's creation. But Keras Layer also has the regularization construction after weights' creation and it almost the same with in the variable_scope. 
To unify the regularizer and reserve the logic in variable_scope.py will benefit `DistributionStrategy`. Because when we create one `MirroredVariable`, the corresponding all regularizations should also be created. We can get them from `GraphKeys.REGULARIZATION_LOSS` and wrap them with `Mirrored` type and replace in the `tf.collections`.

**Any Other info.**
I don't know whether it is suitable to make this change because `Keras` is highly recommended in TF2.0. But the same function in two different places is unnecessary. If you agree with me, I will try to contribute for this.
@lukaszkaiser @fchollet

**Comment by karmel** [comment]
In 2.0, we have removed collections (eg REGULARIZATION_LOSSES) and variable_scope. Regularizers have also been removed from the variable construction APIs, such that they are managed by Keras only. Does that work for you?

**Comment by wangsiyu** [comment]
> In 2.0, we have removed collections (eg REGULARIZATION_LOSSES) and variable_scope. Regularizers have also been removed from the variable construction APIs, such that they are managed by Keras only. Does that work for you?

Thanks for your reply. I think this is a big change. But how to get regularization loss in 2.0 as your description? Before your reply, I think some special resource variables like `MirroredVariable` should manage their regularization themselves.

**Comment by karmel** [comment]
You can use [Keras regularizers](https://www.tensorflow.org/api_docs/python/tf/keras/regularizers) in 2.0. Examples involving regularization:
https://www.tensorflow.org/beta/tutorials/keras/overfit_and_underfit 
https://www.tensorflow.org/beta/guide/keras/training_and_evaluation#handling_losses_and_metrics_that_dont_fit_the_standard_signature
https://www.tensorflow.org/beta/guide/migration_guide#before_converting

**Comment by wangsiyu** [comment]
OK. I want to report a bug here. In Keras, the regularizer's declaration depends on whether the variables exists in global collections. See here :https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/layers/base.py#L421.
But in `MirroredStrategy`, the `MirroredVariable` across all devices is created in the first replica and the other replica will reuse the `MirroredVariable`. That is to say, the variables have been available to other replicas. This will lead to the regularizers are declared only on the first replica, the other replicas will not have any regularizers because of the `existing_variables`. 
Another question is that if you remove global collections, how to  make judgement about whether to declare regularizers？

**Comment by wangsiyu** [comment]
> You can use [Keras regularizers](https://www.tensorflow.org/api_docs/python/tf/keras/regularizers) in 2.0. Examples involving regularization:
> 
> * https://www.tensorflow.org/beta/tutorials/keras/overfit_and_underfit
> * https://www.tensorflow.org/beta/guide/keras/training_and_evaluation#handling_losses_and_metrics_that_dont_fit_the_standard_signature
> * https://www.tensorflow.org/beta/guide/migration_guide#before_converting

OK. I want to report a bug here. In Keras, the regularizer's declaration depends on whether the variables exists in global collections. See here :https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/layers/base.py#L421.
But in `MirroredStrategy`, the `MirroredVariable` across all devices is created in the first replica and the other replica will reuse the `MirroredVariable`. That is to say, the variables have been available to other replicas. This will lead to the regularizers are declared only on the first replica, the other replicas will not have any regularizers because of the `existing_variables`. 
Another question is that if you remove global collections, how to  make judgement about whether to declare regularizers？

**Comment by karmel** [comment]
The code linked to is for v1 tf.layers, not Keras layers, hence the use of global variables. Those layers do not support distribution strategies.

**Comment by wangsiyu** [comment]
> The code linked to is for v1 tf.layers, not Keras layers, hence the use of global variables. Those layers do not support distribution strategies.

I see. Thanks very much.

**Comment by ymodak** [comment]
Closing this issue since its addressed. Feel free to reopen if have any further questions. Thanks!
